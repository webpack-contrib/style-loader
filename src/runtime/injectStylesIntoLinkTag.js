module.exports = (url, options) => {
  options = options || {};
  options.attributes =
    typeof options.attributes === "object" ? options.attributes : {};

  if (typeof options.attributes.nonce === "undefined") {
    const nonce =
      typeof __webpack_nonce__ !== "undefined" ? __webpack_nonce__ : null;

    if (nonce) {
      options.attributes.nonce = nonce;
    }
  }

  const linkElement = document.createElement("link");

  linkElement.rel = "stylesheet";
  linkElement.href = url;

  Object.keys(options.attributes).forEach((key) => {
    linkElement.setAttribute(key, options.attributes[key]);
  });

  options.insert(linkElement);

  return (newUrl) => {
    if (typeof newUrl === "string") {
      linkElement.href = newUrl;
    } else {
      linkElement.parentNode.removeChild(linkElement);
    }
  };
};
