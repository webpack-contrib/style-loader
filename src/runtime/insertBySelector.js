const memo = {};

/* istanbul ignore next  */
function getTarget(target) {
  if (typeof memo[target] === "undefined") {
    let styleTarget = document.querySelector(target);

    // Special case to return head of iframe instead of iframe itself
    if (
      window.HTMLIFrameElement &&
      styleTarget instanceof window.HTMLIFrameElement
    ) {
      try {
        // This will throw an exception if access to iframe is blocked
        // due to cross-origin restrictions
        styleTarget = styleTarget.contentDocument.head;
      } catch (e) {
        // istanbul ignore next
        styleTarget = null;
      }
    }

    memo[target] = styleTarget;
  }

  return memo[target];
}

/* istanbul ignore next  */
function insertBySelector(insert, style) {
  const target = getTarget(insert);

  if (!target) {
    throw new Error(
      "Couldn't find a style target. This probably means that the value for the 'insert' parameter is invalid."
    );
  }

  target.appendChild(style);
}

module.exports = insertBySelector;
