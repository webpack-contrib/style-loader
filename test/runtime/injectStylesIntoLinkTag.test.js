/**
 * @jest-environment jsdom
 */

/* eslint-env browser */

import injectStylesIntoLinkTag from "../../src/runtime/injectStylesIntoLinkTag";

import insertBySelector from "../../src/runtime/insertBySelector";

const getInsertFn = (place) => insertBySelector.bind(null, place);

function insertAtTop(element) {
  const parent = document.querySelector("head");
  // eslint-disable-next-line no-underscore-dangle
  const lastInsertedElement = window._lastElementInsertedByStyleLoader;

  if (!lastInsertedElement) {
    parent.insertBefore(element, parent.firstChild);
  } else if (lastInsertedElement.nextSibling) {
    parent.insertBefore(element, lastInsertedElement.nextSibling);
  } else {
    parent.appendChild(element);
  }

  // eslint-disable-next-line no-underscore-dangle
  window._lastElementInsertedByStyleLoader = element;
}

function insertBeforeAt(element) {
  const parent = document.querySelector("head");
  const target = document.querySelector("#id");

  // eslint-disable-next-line no-underscore-dangle
  const lastInsertedElement = window._lastElementInsertedByStyleLoader;

  if (!lastInsertedElement) {
    parent.insertBefore(element, target);
  } else if (lastInsertedElement.nextSibling) {
    parent.insertBefore(element, lastInsertedElement.nextSibling);
  } else {
    parent.appendChild(element);
  }

  // eslint-disable-next-line no-underscore-dangle
  window._lastElementInsertedByStyleLoader = element;
}

describe("addStyle", () => {
  let defaultOptions;

  beforeEach(() => {
    document.head.innerHTML = "<title>Title</title>";
    document.body.innerHTML = "<h1>Hello world</h1>";

    defaultOptions = {
      insert: getInsertFn("head"),
    };
  });

  // Each query should have be unique because style-loader caching styles in dom

  it("should work", () => {
    injectStylesIntoLinkTag("./style-1.css", defaultOptions);

    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it('should work with "attributes" option', () => {
    injectStylesIntoLinkTag("./style-2.css", {
      ...defaultOptions,
      attributes: { foo: "bar" },
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it('should work with "__webpack_nonce__" variable', () => {
    // eslint-disable-next-line no-underscore-dangle
    window.__webpack_nonce__ = "12345678";

    injectStylesIntoLinkTag("./style-3.css", defaultOptions);

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    // eslint-disable-next-line no-underscore-dangle, no-undefined
    window.__webpack_nonce__ = undefined;
  });

  it('should work with "nonce" attribute and "__webpack_nonce__" variable', () => {
    // eslint-disable-next-line no-underscore-dangle
    window.__webpack_nonce__ = "12345678";

    injectStylesIntoLinkTag("./style-4.css", {
      ...defaultOptions,
      attributes: { nonce: "87654321" },
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    // eslint-disable-next-line no-underscore-dangle, no-undefined
    window.__webpack_nonce__ = undefined;
  });

  it('should work with "insert" option', () => {
    injectStylesIntoLinkTag("./style-5.css", {
      ...defaultOptions,
      insert: getInsertFn("head"),
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it('should work with "insert" option #2', () => {
    injectStylesIntoLinkTag("./style-6.css", {
      ...defaultOptions,
      insert: getInsertFn("body"),
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it('should work with "insert" option #3', () => {
    document.body.innerHTML =
      "<h1>Hello world</h1><iframe class='iframeTarget'/>";

    injectStylesIntoLinkTag("./style-7.css", {
      ...defaultOptions,
      insert: getInsertFn("iframe.iframeTarget"),
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();
    expect(
      document.getElementsByClassName("iframeTarget")[0].contentDocument.head
        .innerHTML
    ).toMatchSnapshot();
  });

  it('should work with "insert" option #4', () => {
    injectStylesIntoLinkTag("./style-8.css", {
      ...defaultOptions,
      insert: insertAtTop,
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    // eslint-disable-next-line no-underscore-dangle
    window._lastElementInsertedByStyleLoader = null;
  });

  it('should work with "insert" option #5', () => {
    document.head.innerHTML =
      '<title>Title</title><script src="https://example.com/script.js" id="id"></script>';

    injectStylesIntoLinkTag("./style-9.css", {
      ...defaultOptions,
      insert: insertBeforeAt,
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    // eslint-disable-next-line no-underscore-dangle
    window._lastElementInsertedByStyleLoader = null;
  });

  it('should throw error with incorrect "insert" option', () => {
    expect(() =>
      injectStylesIntoLinkTag("./style-10.css", {
        ...defaultOptions,
        insert: getInsertFn("invalid"),
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw error with invalid "insert" option', () => {
    expect(() =>
      injectStylesIntoLinkTag("./style-11.css", {
        ...defaultOptions,
        insert: getInsertFn("#test><><><"),
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it("should work with updates", () => {
    const update = injectStylesIntoLinkTag("./style-12.css", defaultOptions);

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    update("./style-12.css");

    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it("should work with updates #2", () => {
    const update = injectStylesIntoLinkTag("./style-13.css", defaultOptions);

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    update("./style-14.css");

    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it("should work with updates #3", () => {
    const update = injectStylesIntoLinkTag("./style-15.css", defaultOptions);

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    update();

    expect(document.documentElement.innerHTML).toMatchSnapshot();
  });

  it("should work with updates #4", () => {
    const update = injectStylesIntoLinkTag("./style-16.css", {
      ...defaultOptions,
      insert: insertAtTop,
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    update("./style-16.css");

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    // eslint-disable-next-line no-underscore-dangle
    window._lastElementInsertedByStyleLoader = null;
  });

  it("should work with updates #5", () => {
    document.head.innerHTML =
      '<title>Title</title><script src="https://example.com/script.js" id="id"></script>';

    const update = injectStylesIntoLinkTag("./style-17.css", {
      ...defaultOptions,
      insert: insertBeforeAt,
    });

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    update("./style-17.css");

    expect(document.documentElement.innerHTML).toMatchSnapshot();

    // eslint-disable-next-line no-underscore-dangle
    window._lastElementInsertedByStyleLoader = null;
  });
});
