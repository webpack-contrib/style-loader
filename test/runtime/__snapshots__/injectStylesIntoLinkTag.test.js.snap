// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addStyle should throw error with incorrect "insert" option 1`] = `"Couldn't find a style target. This probably means that the value for the 'insert' parameter is invalid."`;

exports[`addStyle should throw error with invalid "insert" option 1`] = `"'#test><><><' is not a valid selector"`;

exports[`addStyle should work 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-1.css\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with "__webpack_nonce__" variable 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-3.css\\" nonce=\\"12345678\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with "attributes" option 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-2.css\\" foo=\\"bar\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with "insert" option #2 1`] = `"<head><title>Title</title></head><body><h1>Hello world</h1><link rel=\\"stylesheet\\" href=\\"./style-6.css\\"></body>"`;

exports[`addStyle should work with "insert" option #3 1`] = `"<head><title>Title</title></head><body><h1>Hello world</h1><iframe class=\\"iframeTarget\\"></iframe></body>"`;

exports[`addStyle should work with "insert" option #3 2`] = `"<link rel=\\"stylesheet\\" href=\\"./style-7.css\\">"`;

exports[`addStyle should work with "insert" option #4 1`] = `"<head><link rel=\\"stylesheet\\" href=\\"./style-8.css\\"><title>Title</title></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with "insert" option #5 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-9.css\\"><script src=\\"https://example.com/script.js\\" id=\\"id\\"></script></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with "insert" option 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-5.css\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with "nonce" attribute and "__webpack_nonce__" variable 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-4.css\\" nonce=\\"87654321\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #2 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-13.css\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #2 2`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-14.css\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #3 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-15.css\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #3 2`] = `"<head><title>Title</title></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #4 1`] = `"<head><link rel=\\"stylesheet\\" href=\\"./style-16.css\\"><title>Title</title></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #4 2`] = `"<head><link rel=\\"stylesheet\\" href=\\"./style-16.css\\"><title>Title</title></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #5 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-17.css\\"><script src=\\"https://example.com/script.js\\" id=\\"id\\"></script></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates #5 2`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-17.css\\"><script src=\\"https://example.com/script.js\\" id=\\"id\\"></script></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates 1`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-12.css\\"></head><body><h1>Hello world</h1></body>"`;

exports[`addStyle should work with updates 2`] = `"<head><title>Title</title><link rel=\\"stylesheet\\" href=\\"./style-12.css\\"></head><body><h1>Hello world</h1></body>"`;
